Add crosshair support for Android

srifqi opened the pull request:
If `touch_use_crosshair` is enabled, a crosshair will be shown to select the object. This will give Android (or any touch screen interface) player a way to play like they play on desktop (with mouse).

For third-person camera mode, it is same as the desktop version: on third-person back camera mode, player is forced to use crosshair; while on third-person front camera mode, player is unable to select anything.

Possibly fixes #7645 and fixes #6181.

[YouTube video (1)](https://youtu.be/wF1Dpdrnhds)
[YouTube video (2)](https://youtu.be/Hw73Ohyodls)

stujones11 commented:
I welcome this option, however, it only really fixes #6181 if this is enforced it third person view. I don't see any code that is doing this.

srifqi commented:
> I do welcome this as an option, however, it only really fixes #6181 if this is enforced in third person view. I don't see any code that is doing this.

Just added. Thanks for remind me.

stujones11 commented:
Tested and working. :+1: 

The code also seems fine except maybe some questionable styling, however this file clearly must be whitelisted, anyway :D

waterel commented:
When you slide your finger while digging, does the camera rotate? If not, this pull request doesn't fix issue #7645. The YouTube video doesn't show what the case is.

nerzhul commented on src/gui/touchscreengui.cpp:
```diff
@@ -1256,6 +1284,11 @@ void TouchScreenGUI::registerHudItem(int index, const rect<s32> &rect)
 	m_hud_rects[index] = rect;
 }
 
+void TouchScreenGUI::setCameraMode(CameraMode camera_mode)
```
this function can be inlined

srifqi replied:
Okay.

srifqi commented:
> When you slide your finger while digging, does the camera rotate? If not, this pull request doesn't fix issue #7645. The YouTube video doesn't show what the case is.

Yes. The camera still rotates and the player is still digging. Watch [my new video](https://youtu.be/Hw73Ohyodls).

srifqi commented:
I just realised that something doesn't seems right on third-person back camera mode. You can't select as far as when playing on desktop.
@stujones11, can you try please?

stujones11 commented:
@srifqi I have noticed that crosshair placement is rather awkward on my phone due to its very wide aspect screen, although my tablet seems much the windowed desktop version.

The only other reason I can think why placement distance might be different for android would be this commit: 2c450ed93f0217d4950bcd4d2be3d5d1b92d7a2d Are you comparing this in creative or survival mode?

srifqi commented:
I'm still searching for camera offset thing-y. Maybe, you have ideas where to search?

> Are you comparing this in creative or survival mode?

Yes. It has lower distance in survival mode. I cannot find any line that make it differs. I just tested and selection distance bug also happens on desktop.

ClobberXD commented on src/client/game.cpp:
```diff
@@ -3802,7 +3804,10 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats, f32 dtime,
 	try {
-		draw_crosshair = !g_settings->getBool("touchtarget");
+		draw_crosshair = !g_settings->getBool("touchtarget") ||
+				(g_settings->getBool("use_crosshair") &&
```
Isn't it better to locally cache these settings, instead of retrieving them every frame? I assume `Game::updateFrame` is run once every frame.

srifqi replied:
I agree.

ClobberXD commented on src/gui/touchscreengui.cpp:
```diff
@@ -761,8 +762,13 @@ void TouchScreenGUI::handleReleaseEvent(int evt_id)
+				translated->MouseInput.Y            = m_screensize.Y / 2;
+			} else {
+				translated->MouseInput.X            = m_move_downlocation.X;
+				translated->MouseInput.Y            = m_move_downlocation.Y;
```
Incorrect alignment. Also, now that these lines are inside if-else blocks, the `=` needn't be aligned with the lines outside the blocks.

srifqi replied:
Done. I also moved it to bottom.

ClobberXD commented on src/gui/touchscreengui.cpp:
```diff
@@ -886,7 +892,11 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
+						m_move_downlocation = v2s32(m_screensize.X / 2, m_screensize.Y / 2);
+					} else {
+						m_move_downlocation = v2s32(event.TouchInput.X, event.TouchInput.Y);
+					}
```
Moving the two if-else blocks out of the way of these aligned lines would make the code look nicer. :)

srifqi replied:
Done.

ClobberXD commented on src/gui/touchscreengui.cpp:
```diff
@@ -909,7 +919,8 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
 			if ((event.TouchInput.ID == m_move_id) &&
-				(!m_move_sent_as_mouse_event)) {
+				(!m_move_sent_as_mouse_event || m_use_crosshair ||
+				m_camera_mode == CAMERA_MODE_THIRD)) {
```
These two lines need to be indented by one more tab to not align with the code within the if block

srifqi replied:
Done.

ClobberXD commented on src/gui/touchscreengui.cpp:
```diff
@@ -909,7 +919,8 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
 
 		if (m_move_id != -1) {
 			if ((event.TouchInput.ID == m_move_id) &&
```
This line hasn't been touched, but nevertheless

```suggestion
			if (event.TouchInput.ID == m_move_id &&
```

srifqi replied:
Done.

sfan5 commented:
Isn't the touchtarget setting supposed to enable/disable the crosshair?
What happens if you use `touchtarget = false` and `use_crosshair = false`?

ClobberXD commented on src/gui/touchscreengui.cpp:
```diff
@@ -892,12 +892,12 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
 					m_move_has_really_moved    = false;
 					m_move_downtime            = porting::getTimeMs();
+					m_move_sent_as_mouse_event = false;
 					if (m_use_crosshair || m_camera_mode == CAMERA_MODE_THIRD) {
```
Braces aren't required as there's only one line of code within the if and the else blocks.

srifqi replied:
Should I also fix for other lines?

ClobberXD replied:
It'd be great if you could fix this for all touched lines.

srifqi replied:
Is 8ec9524e456bcaa73b2d79f74ccdb11bfa7bb8b0 enough? I can't find other touched lines that should be fixed. Did you mean also neighbouring lines?

ClobberXD replied:
I didn't check whether there were any other lines that needed fixing; I assumed there were, when you said "Should I also fix this for other lines?". But looking at the code, I don't see any other line that adds braces to single-line blocks. So please ignore my last request. :)

srifqi commented:
> Isn't the touchtarget setting supposed to enable/disable the crosshair?

Yes, I guess. I just realised that `touchtarget` setting is not used anymore, except to show/hide the crosshair. This setting even is not in `settingtypes.txt`, but is in `defaultsettings.cpp`. Maybe, we should remove this setting completely?

> What happens if you use `touchtarget = false` and `use_crosshair = false`?

I have not tested it. But, I think, it will show the crosshair as `touchtarget` enables the crosshair when is set to `false`.
```cpp
bool draw_crosshair = !m_android_touchtarget ||
		(m_android_use_crosshair &&
		camera->getCameraMode() != CAMERA_MODE_THIRD_FRONT) ||
		camera->getCameraMode() == CAMERA_MODE_THIRD;
```

srifqi commented:
I [rebased](http://irc.minetest.net/minetest-dev/2020-01-30#i_5636790) this PR and added [a fix](https://github.com/minetest/minetest/pull/7865/commits/cc8f2ec9b56360f996b06536ff64ec27290d4ad0) for wrong shootline distance calculation in third-person rear view (almost same as #8261).

ronoaldo commented:
I would like to help any fix/testing/qa for this. How can I help?

srifqi commented:
@ronoaldo, thank you for your willingness in testing this PR. I am sorry for being late in replying. I have rebased this PR to `minetest/master` and tested it locally.

Here are the APKs for testing:
- (arm64-v8a) https://drive.google.com/file/d/1Eb9Vj_jZiGrf78CTlUdP2Y1HtfRWHnOJ/view
- (armeabi-v7a) https://drive.google.com/file/d/1zaD39dsAtsD3-v4XjefNYDAPO8QfH6Kg/view

(Those APKs were compiled using NDK 22.0.7026061.)

We need to test these:
- When `use_crosshair` is enabled, there is a crosshair at the center of the screen.
- In third-person back camera mode, there is a crosshair and the player only able to point using crosshair.
- Player can not interact in third-person front camera mode.

Thank you!

rollerozxa commented:
Any updates on this?

sfan5 commented:
It still needs a rebase. If you are interest you can consider adopting the PR.

srifqi commented:
I am sorry for my late response. I have rebased and squashed this PR.

Here are the APKs for testing:
- (arm64-v8a) https://drive.google.com/file/d/1QdT4XiJOlFU5TMmRC_74rGWKrdaZA-0-/view
- (armeabi-v7a) https://drive.google.com/file/d/1j_u_xgXUcxbMkGcB5nq9sMRW2hZ4A7kk/view

Those APKs were compiled using NDK 23.1.7779620 (more recent than `master` 23.0.7599858).

LoneWolfHT commented:
Tested, works.
It would be nice if I could make the crosshair bigger but I think that's best left for another PR

rollerozxa commented:
Tested, works well. 

I'd even say that this should be enabled by default, because this literally makes Minetest Android controls usable. How hasn't this been merged yet?

rollerozxa commented:
Compiling fails when doing a desktop build with HAVE_TOUCHSCREENGUI=1, [here's a fix for it](https://github.com/rollerozxa/rollertest/commit/cee84b23319376a7d78a8e447c6b14e5e415d14f)

srifqi commented:
Thank you for the fix, @rollerozxa.

I have cherry-picked it and rebased this PR.

rubenwardy commented:
Bug: if you walk without rotating, the selected node doesn't change - it should update to the node under the crosshair

rubenwardy commented:
Fixed

sfan5 commented on builtin/settingtypes.txt:
```diff
@@ -125,6 +125,10 @@ fixed_virtual_joystick (Fixed virtual joystick) bool false
 
+#    (Android) Use crosshair to select object instead of whole screen.
+#    If enabled, a crosshair will be shown and will be used for selecting object.
+use_crosshair (Use crosshair) bool false
```
this should probably be `touch_use_crosshair` if it's specific to that

srifqi replied:
I agree.

sfan5 commented on src/client/game.cpp:
```diff
@@ -3076,16 +3088,16 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud)
-
-	if ((g_settings->getBool("touchtarget")) && (g_touchscreengui)) {
+	if (g_settings->getBool("touchtarget") && g_touchscreengui &&
+			!g_settings->getBool("use_crosshair") &&
```
Use the `m_android_*` variables here?

srifqi replied:
Yes.

sfan5 commented on src/client/game.cpp:
```diff
@@ -3960,14 +3972,15 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats, f32 dtime,
+	bool draw_crosshair = !m_android_touchtarget ||
+			(m_android_use_crosshair &&
+			camera->getCameraMode() != CAMERA_MODE_THIRD_FRONT) ||
+			camera->getCameraMode() == CAMERA_MODE_THIRD;
```
this totally overwrites the normal visibility calculations and is also subtly different?!
a variation of the following would be better:
```cpp
    if (m_android_touchtarget)
        draw_crosshair = false;
```
...if use_crosshair can be fit into it that is, I don't entirely get how that works

srifqi replied:
I have rewritten it to be same as non-touch, except when in the first-person camera mode.

sfan5 commented on src/gui/touchscreengui.cpp:
```diff
@@ -678,12 +679,17 @@ void TouchScreenGUI::handleReleaseEvent(size_t evt_id)
 			translated->MouseInput.Control      = false;
 			translated->MouseInput.ButtonStates = 0;
 			translated->MouseInput.Event        = EMIE_LMOUSE_LEFT_UP;
+			if (m_use_crosshair || m_camera_mode == CAMERA_MODE_THIRD) {
```
Instead of replicating the condition here (error prone) and even passing `m_camera_mode` around why not directly pass the current state of `draw_crosshair` to touchscreengui?

srifqi replied:
Yes, `m_draw_crosshair` will be used. The calculation is moved into `game.cpp` (`touchscreengui.cpp` only receive the result).

sfan5 commented on src/gui/touchscreengui.cpp:
```diff
@@ -1010,11 +1023,18 @@ bool TouchScreenGUI::doubleTapDetection()
 	if (distance > (20 + m_touchscreen_threshold))
 		return false;
 
+	s32 mX = m_key_events[0].x;
+	s32 mY = m_key_events[0].y;
```
v2s32

srifqi replied:
Done

srifqi commented:
I changed some things from the reviews:
- The crosshair calculation (in determining enable/disable) is moved into `game.cpp`.
- The setting `use_crosshair` is now `touch_use_crosshair`.
- The "touch target" is removed completely.
- The crosshair support is not just for Android, but also for all that uses touch screen. (This is already done since rollerozxa's commit but I explain it more in `settingtypes.txt` etc.)

x2048 commented on src/gui/touchscreengui.cpp:
```diff
@@ -849,11 +855,14 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
 					m_camera_pitch = MYMIN(MYMAX(m_camera_pitch + (dy * d), -180), 180);
 
 					// update shootline
+					if (m_draw_crosshair) {
```
It looks like m_shootline is not use at all if m_draw_crosshair is true. Is it necessary to make sure it has a correct value?

rubenwardy replied:
The shootline is updated elsewhere, in the same place desktop does it

srifqi replied:
I do not know. I wrote it just for consistency with other code that is there. I am guessing that it is not necessary for this very case.

Zughy commented:
@srifqi please have a look at 2048's review, just one appproval missing :)

srifqi commented on src/client/game.cpp:
```diff
@@ -3970,10 +3980,8 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats, f32 dtime,
-		draw_crosshair = !g_settings->getBool("touchtarget");
-	} catch (SettingNotFoundException) {
-	}
+	if (!m_touch_use_crosshair && camera->getCameraMode() == CAMERA_MODE_FIRST)
```
This value (`!m_touch_use_crosshair && camera->getCameraMode() == CAMERA_MODE_FIRST`) is just repeated twice (line 3097 and 3996. Should it be a new variable/method? If required, I am thinking about `isNoCrosshairAllowed()` for its name.

srifqi commented:
I followed up the review and rebased the PR.

x2048 commented on src/client/game.cpp:
```diff
@@ -2973,6 +2981,12 @@ void Game::updateCamera(f32 dtime)
+#ifdef HAVE_TOUCHSCREENGUI
+		if (g_touchscreengui)
+			g_touchscreengui->setUseCrosshair(m_touch_use_crosshair ||
+					camera->getCameraMode() == CAMERA_MODE_THIRD);
```
I would argue this is functionally equivalent to !isNoCrosshairAllowed() since front 3rd person disallows any interaction with the world.

srifqi replied:
That is correct. Thank you for pointing it out!

https://github.com/minetest/minetest/blob/0251b01da66a67ed84f8d27c33aa35f34aaca237/src/client/game.cpp#L3085-L3089

https://github.com/minetest/minetest/blob/0251b01da66a67ed84f8d27c33aa35f34aaca237/src/client/game.cpp#L3990-L3992

x2048 reviewed (approved):
Looks good.

srifqi commented:
Thank you for the merge!
